raiden_contracts/data/source/services/MonitoringService.sol

```diff
@@ -175,6 +175,15 @@ contract MonitoringService is Utils {
         canMonitor(msg.sender)
         public
     {
+        TokenNetwork token_network = TokenNetwork(token_network_address);
+        uint256 channel_identifier = token_network.getChannelIdentifier(
+            closing_participant, non_closing_participant
+        );
```

Comment:
I have to think through. What can happen when `getChannelIdentifier()` calls back into the MS contract?  (The address `token_network` is given from outside so cannot be trusted.)

Reply:
My current thoughts on this are

1. The client can't be cheated by providing a bad token network, since the client's signature is checked and includes the token network address.

2. The MS could already be cheated before, because it already has a call to `token_network.updateNonClosingBalanceProof`. The MS will have to verify that the token network is registered in the token network registry.

Is it safe to assume that a registered token network behaves as intended in the long run?

Reply:
When `token_network` is malicious, it can call back `monitor()` and register a balance proof of the same nonce on the same channel. For that,
* the channel participant must have given a signed balance proof to the attacker
* the victim MS must have overlooked the weird feature of the TokenNetwork.

What else...?

Reply:
Instead of thinking about all possibilities, I suggest moving everything below `updateReward()`.

* That doesn't change the functionality in the normal executions.  When any of the `require` fails, the storage-change in `updateReward()` is reverted.
* That causes the reentrant call to fail, when it's about the same channel and the same nonse.

Reply:
That helps against reentrancy, but the attacker could still make the transaction fail so that the MS spent gas on the transaction but does not get any reward.

Reply:
> the attacker could still make the transaction fail 

I agree. It's just a matter of making sure nothing worse happens.

Even with the current code, that might be the case, but with reentrancy, the property is harder to establish.  All users need to think about every possible re-entry into any external/public functions.